### PR TITLE
preferences component refactor

### PR DESCRIPTION
**Pull request by no-stack-dub-sack**
An attempt at making the code in the preferences page component a bit "cleaner", i.e. more readable, separating out functions so that each only does one thing, better commenting, better naming etc. 

A quick (and incomplete) summary:
- improved validations code
    - decoupled validations from component, now as helper object, to help eliminate redundancies and to make code cleaner
    - converted ugly if/else blocks into switch statements for improved readability
    - created `isFieldValid` function which separates out code that was being handled by `isSectionValid`
     - eliminated other redundancies such as set errors state expression which was being repeated at the end of every validator function
- split profile strength logic into own function, instead of as part of `isPageValid`
- organized functions into sections - handleChange, validations, helpers, etc.

Best way to look at these might be to just skip the dif and go straight to the raw, for the preferences page, since so much of it was re-written / changed

### CODE REVIEW

File: client/src/components/dashboard/Preferences.js [hunk 1]
```diff
@@ -267,132 +248,225 @@ class Preferences extends React.Component {
...
+        viewState.showSocial = true;
+      if (errors.career)
+        viewState.showCareer = true;
+      if (errors.mentorshipSkills)
+        viewState.showMentorship = true;
+      this.setState({ viewState });
+    });
+  }
 
+  profileStrenth = () => {
```
Comment (bonham000) [inline client/src/components/dashboard/Preferences.js]: This guy is misspelled.

File: client/src/components/dashboard/Preferences.js [hunk 2]
```diff
@@ -17,14 +17,14 @@ import propTypes from 'prop-types';
...
-import skills from '../../assets/dropdowns/skills';
+import skills, { SKILLS_MAP } from '../../assets/dropdowns/skills';
 import SkillsAndInterests from './Profile/Preferences/SkillsAndInterests';
 import Social from './Profile/Preferences/Social';
 import styled from 'styled-components';
 import swearjar from '../../assets/helpers/swearjar-lite';
 import { ThickPaddedBottom } from '../../styles/style-utils';
 import UserLabel from '../dashboard/common/UserLabel';
-import Validator from 'validator';
+import __ from '../../assets/helpers/validations';
```
Comment (bonham000) [inline client/src/components/dashboard/Preferences.js]: I would probably name this something more descriptive.

File: client/src/components/dashboard/Preferences.js [hunk 3]
```diff
@@ -118,31 +119,31 @@ class Preferences extends React.Component {
...
+  HANDLE CHANGE FUNCTIONS
+  **********************/
 
   handleSkillsChange = (e, { value }) => {
     var { user } = this.state;
-    user.skillsAndInterests.coreSkills = value;
+    user.skillsAndInterests.coreSkills = value
+      /* extra logic here and in sister function
+      to make sure same skills with different cases
+      are not added to the list */
```
Comment (bonham000) [inline client/src/components/dashboard/Preferences.js]: Having this comment break the function chaining makes it harder to follow what's happening and is somewhat unusual, I'd probably move the comment move the comment up a few lines.

File: client/src/components/dashboard/Preferences.js [hunk 4]
```diff
@@ -267,132 +248,225 @@ class Preferences extends React.Component {
       updateUserPartial(_id, section, user[section]).then(res => {
         const { updatedUser } = res.data;
         this.props.saveUser(updatedUser);
-        // open "Saved" popup
+        // open "saved" popup
         this.setState({ [e.target.id]: true });
         // close popup 1 second later
         setTimeout( _ => this.resetPopUp(e.target.id), 1200);
```
Comment (bonham000) [inline client/src/components/dashboard/Preferences.js]: I remember this syntax showing up some other time and we talked a little about it... I'm not sure where it became common to write a timeout like this? But if I'm not wrong the `_` is an argument to the function now. In my view it would be better to just write `setTimeout(() => ... ), delay)`. But maybe `_` is like some standard, idk.

File: client/src/components/dashboard/Preferences.js [hunk 5]
```diff
@@ -267,132 +248,225 @@ class Preferences extends React.Component {
...
     }
+    return this.setErrors(errors);
+  }
 
-    if (isEmpty(errors)) {
+  isSectionValid = (section, str) => {
+    var errors = {};
+    const { user } = this.state;
+    const { email } = this.state.user.personal;
+    const { personal, social } = this.state.user;
```
Comment (bonham000) [inline client/src/components/dashboard/Preferences.js]: You could destructure these from the `user` you destructure in line 322.

File: client/src/components/dashboard/Preferences.js [hunk 6]
```diff
@@ -75,13 +75,14 @@ class Preferences extends React.Component {
...
-    this.CODEPEN_ERROR = "Please enter your username only, not your profile url.";
-    this.MENTORSHIP_ERROR = "To complete your mentorship prorgram enrollment, please fill out the section above.";
+    // SHARED VALIDATION ERRORS:
+    this.EMAIL_ERROR = 'Please enter a valid email address.';
+    this.LOCATION_ERROR = 'Location must be 25 characters or less.';
+    this.DISPLAY_NAME_ERROR = 'Display name must be 40 characters or less.';
+    this.CAREER_ERROR = 'Please complete the entire section or clear the form.';
+    this.CODEPEN_ERROR = 'Please enter your username only, not your profile url.';
+    this.COUNTRY_ERROR = 'Please select your country (so we can make a cool D3 map!).';
+    this.MENTORSHIP_ERROR = 'To complete your mentorship prorgram enrollment, please complete the section above.';
```
Comment (bonham000) [inline client/src/components/dashboard/Preferences.js]: These are constants, it might be better to extract them out of the component into a different location.

**Comment by bonham000** [comment]
@no-stack-dub-sack this is looking pretty good, it's hard to believe how complex this component is. Basically, I think it's just really on the line where a lot of logic needs to get abstracted out of it into other files. It's really complex right now and hard to follow in detail what's going on by just glancing through the code.

But since it's probably not getting more complex in the immediate future I think this is all fine for now. I don't really have any major comments other than the comments I pointed out here. Looks good to me!

File: client/src/components/dashboard/Preferences.js (hunk 1 above)
Comment (no-stack-dub-sack) [inline client/src/components/dashboard/Preferences.js]: @bonham000 lol

File: client/src/components/dashboard/Preferences.js (hunk 5 above)
Comment (no-stack-dub-sack) [inline client/src/components/dashboard/Preferences.js]: I could... and probably should... and did... but, I hate the way it looks.

File: client/src/components/dashboard/Preferences.js (hunk 4 above)
Comment (no-stack-dub-sack) [inline client/src/components/dashboard/Preferences.js]: That makes sense...

File: client/src/components/dashboard/Preferences.js (hunk 3 above)
Comment (no-stack-dub-sack) [inline client/src/components/dashboard/Preferences.js]: roger that

File: client/src/components/dashboard/Preferences.js (hunk 6 above)
Comment (no-stack-dub-sack) [inline client/src/components/dashboard/Preferences.js]: cool

File: client/src/components/dashboard/Preferences.js (hunk 2 above)
Comment (no-stack-dub-sack) [inline client/src/components/dashboard/Preferences.js]: @bonham000 could... but the methods on the object are descriptive to the point where it's clear what it is, no? Again, I made this choice based on my stylistic preferences. So maybe that would be annoying to other developers. But also `validator.validateMentorship` is redundant. 

As I was typing this, I just remembered another convention from "Clean Code Javascript"...

I'll rename the variable `validate` and then each of the methods can just be the things that they're validating, e.g. `validate.mentorship`. This will solve the stylistic issues that annoyed me with a long object name plus long method name, as well as the redundancy issues.

Comment (no-stack-dub-sack) [inline client/src/components/dashboard/Preferences.js]: @bonham000 I just went with the first thing before I read this... does that make it more clear like you are saying?

Comment (bonham000) [inline client/src/components/dashboard/Preferences.js]: @no-stack-dub-sack I think either way is OK, just different options.

**Comment by no-stack-dub-sack** [comment]
@bonham000 made those improvements... 

# Squash it...

you can delete all the commit details from the squash and just call it something like "preferences component refactor" or something along those lines if you end up merging. Won't matter much if you do or don't &mdash; I'm away this weekend and won't get to redploy until Sunday anyway.
